Clement/eng 2181 otel exporter

clementsirieix opened the pull request:
## Changes

- Deprecate instrumentations call
- Added `initialize` method to the client
- Added a `set_properties` method for customization
- Added a custom Traceloop exporter to handle llm spans

## How to test?

### Test deprecations

- Test any examples that use an explicit instrumentation, notice the deprecation log

### Tests

- The basic streaming test `LITERAL_API_URL="http://localhost:3000" LITERAL_API_KEY="my-initial-api-key" python3 examples/streaming.py`

<img width="1127" alt="Screenshot 2024-12-16 at 17 45 55" src="https://github.com/user-attachments/assets/cdb2cfb2-6efb-4d51-8e68-ecaf42cab50e" />

- The langchain variables example `LITERAL_API_URL="http://localhost:3000" LITERAL_API_KEY="my-initial-api-key" python3 examples/langchain_variable.py`

<img width="1129" alt="Screenshot 2024-12-16 at 17 46 13" src="https://github.com/user-attachments/assets/c63acfec-0694-40b7-9965-de349a169b6e" />

- Test multimodal query `LITERAL_API_URL="http://localhost:3000" LITERAL_API_KEY="my-initial-api-key" python3 examples/multimodal.py`

<img width="1131" alt="Screenshot 2024-12-16 at 17 45 30" src="https://github.com/user-attachments/assets/a1e45098-ba20-444c-b82c-10992cadb487" />

- Test llamaindex flow `LITERAL_API_URL="http://localhost:3000" LITERAL_API_KEY="my-initial-api-key" python3 examples/llamaindex_workflow.py`

![Screenshot 2024-12-17 at 17 41 54](https://github.com/user-attachments/assets/07c5bfd7-53cd-4dbf-8a6b-ddd6ebf332fe)

clementsirieix commented on literalai/client.py:
```diff
@@ -119,6 +128,13 @@ def instrument_llamaindex(self):
         instrument_llamaindex(self.to_sync())
 
+    def initialize(self):
+        with redirect_stdout(io.StringIO()):
```
Traceloop is verbose so this is aimed to prevent explicit `print` statement.

clementsirieix commented on literalai/client.py:
```diff
@@ -352,6 +368,27 @@ def get_current_root_run(self):
+        root_run = active_root_run_var.get()
+        parent = active_steps_var.get()[-1] if active_steps_var.get() else None
+
+        Traceloop.set_association_properties(
```
This is not an upsert so we need to maintain explicit calls.

clementsirieix commented on literalai/client.py:
```diff
@@ -352,6 +368,27 @@ def get_current_root_run(self):
+
+        Traceloop.set_association_properties(
+            {
+                "literal.thread_id": str(thread.id) if thread else "None",
```
Can only store strings

clementsirieix commented on literalai/event_processor.py:
```diff
@@ -98,6 +98,7 @@ def _process_batch(self, batch: List):
             self.processing_counter -= len(batch)
 
     def flush_and_stop(self):
+        time.sleep(4)
```
Found an issue where the llm step is added to the queue but not flushed, would like to solve this.

clementsirieix commented on literalai/exporter.py:
```diff
@@ -0,0 +1,443 @@
+        """Force flush the exporter."""
+        return True
+
+    #     # TODO: Add generation promptid
+    #     # TODO: Add generation variables
+    #     # TODO: Check missing variables
+    #     # TODO: ttFirstToken
+    #     # TODO: duration
+    #     # TODO: tokenThroughputInSeconds
+    #     # TODO: Add tools
+    #     # TODO: error check with gemini error
```
Haven't found a proper way to handle those as of now:
- promptid and variables could be set manually?
- missing variables like `frequency_penalty` are just not tracked by Traceloop so they do not appear in the span attributes, could only found basic support in their issue.
- ttFirstToken, duration and tokenThroughputInSeconds could not find a clear alternative
- tools to test

clementsirieix replied:
Could not find a way to handle gemini and ttFirstToken

willydouhard commented on literalai/observability/step.py:
```diff
@@ -549,6 +566,21 @@ def __enter__(self) -> Step:
 
         if active_root_run_var.get() is None and self.step_type == "run":
             active_root_run_var.set(self.step)
+            Traceloop.set_association_properties(
```
if otel is not initialized this wont be an issue?

clementsirieix replied:
Yes it can be safely called even without initialization 👍

clementsirieix commented:
As discussed yesterday:
- Added a static step to the llamaindex workflow example
- Updated the 2 todos pointing towards the github work:
  - PR opened to trace llamaindex workflows https://github.com/traceloop/openllmetry/pull/2421
  - Gemini issue https://github.com/traceloop/openllmetry/issues/2419
